### PR TITLE
Liaye/bump typescript to3 0 1

### PR DESCRIPTION
**Pull request by ipip2005**

### CODE REVIEW

File: apps/rush-lib/tsconfig.json [hunk 1]
```diff
@@ -3,7 +3,8 @@
   "compilerOptions": {
     "types": [
       "jest"
-    ]
+    ],
+    "skipLibCheck": true
```
Comment (ipip2005) [inline apps/rush-lib/tsconfig.json]: The dependency @types/jest build fail in this project. Add this flag to avoid building that file

**Comment by pgonzal** [comment]
@ipip2005 thanks for creating this PR.  However, upgrading the TypeScript compiler in **gulp-core-build-typescript** will require all the projects at Microsoft that use it to migrate to the new compiler also.  We want to do this, but historically it has always required about a week of work.  For example, in TypeScript 3.x the `Request` interface has a new property, which requires classes implementing this interface to have the new property, and these are third-party public APIs that need an API review and documentation step.  We have also seen runtime regressions from compiler upgrades, so manual testing is required.  (These issues typically don't affect small projects much at all, but they are significant in our large internal monorepos.)

Bumping the major version of **gulp-core-build-typescript** seems like the way to allow people to "opt-in" to the upgrade, but unfortunately the **gulp-core-build-____** packages are coupled such that all entangled packages (including internal build rigs) would need to be bumped together.  That may be difficult to manage.

Instead, we've been planning to handle this upgrade a different way:  We want to move the compiler-related tasks into **rush-stack-compiler**, and isolate the TypeScript dependency there.  Then we will make **rush-stack-compiler** a peer dependency that allows projects to select a (supported) compiler version by choosing their version for **rush-stack-compiler**.  ([Rush Stack](https://www.npmjs.com/package/@microsoft/rush-stack) is a larger effort to untangle the coupling that results from gulp pipelines.)

+ @iclanton who was investigating this

**Comment by pgonzal** [comment]
That said, being TypeScript 2.4.1 is ancient and we really need to upgrade.  Thanks for helping call attention to this.  :-)

**Comment by ipip2005** [comment]
@pgonzal Thanks for the info. Looking forward to good news! Your planned approach looks promising.

**Pull request closed without merging**